src/main/scala/io/treeverse/clients/ApiClient.scala

```diff
@@ -1,25 +1,40 @@
 package io.treeverse.clients
 
+import com.google.common.cache.CacheBuilder
 import org.json4s._
 import org.json4s.native.JsonMethods._
 import scalaj.http.Http
 
-class ApiClient(apiUrl : String, accessKey : String, secretKey : String) {
-  def getRepositoryStorageNamespace(repoName: String): String = {
-    val resp = Http("%s/repositories/%s".format(apiUrl, repoName)).header("Accept", "application/json").auth(accessKey, secretKey).asString
-    val JString(storageNamespace) = parse(resp.body) \ "storage_namespace"
-    storageNamespace.replace("s3://", "s3a://") // TODO(johnnyaug) solve this without replace
+import java.net.URI
+import java.time.Duration
+
+class ApiClient(apiUrl: String, accessKey: String, secretKey: String) {
+  private val cache = CacheBuilder.newBuilder().expireAfterWrite(Duration.ofMinutes(2)).build[String, String]()
+
+  private def getStorageNamespace(repoName: String): String = {
+    cache.get(repoName, () => {
+      val getRepositoryURI = URI.create("%s/repositories/%s".format(apiUrl, repoName)).normalize()
+      val resp = Http(getRepositoryURI.toString).header("Accept", "application/json").auth(accessKey, secretKey).asString
+      val JString(storageNamespace) = parse(resp.body) \ "storage_namespace"
+      var snUri = URI.create(storageNamespace)
+      snUri = new URI(if (snUri.getScheme == "s3") "s3a" else snUri.getScheme, snUri.getHost, snUri.getPath, snUri.getFragment) // TODO(johnnyaug) solve this without replace
+      snUri.normalize().toString
+    })
   }
 
-  def getMetaRangeURL(repoName: String, storageNamespace: String, commitID: String): String = {
-    var resp = Http("%s/repositories/%s/commits/%s".format(apiUrl, repoName, commitID)).header("Accept", "application/json").auth(accessKey, secretKey).asString
+  def getMetaRangeURL(repoName: String, commitID: String): String = {
+    val getCommitURI = URI.create("%s/repositories/%s/commits/%s".format(apiUrl, repoName, commitID)).normalize()
+    println(getCommitURI)
+    var resp = Http(getCommitURI.toString).header("Accept", "application/json").auth(accessKey, secretKey).asString
     val JString(metaRangeID) = parse(resp.body) \ "meta_range_id"
-    resp = Http("%s/repositories/%s/metadata/meta_range/%s".format(apiUrl, repoName, metaRangeID)).header("Accept", "application/json").auth(accessKey, secretKey).asString
-    storageNamespace + "/" + resp.header("Location").get
+    val getMetaRangeURI = URI.create("%s/repositories/%s/metadata/meta_range/%s".format(apiUrl, repoName, metaRangeID)).normalize()
+    resp = Http(getMetaRangeURI.toString).header("Accept", "application/json").auth(accessKey, secretKey).asString
+    URI.create(getStorageNamespace(repoName) + "/" + resp.header("Location").get).normalize().toString
```

Comment:
I prefer to use `resolve` rather than trying to glue URIs, it is usually less fragile.  I think `getStorageNamespace` should cache and return a URI (and maybe change its name).  Then you can write
```suggestion
    getStorageNamespace(repoName).resolve(resp.header("Location").get).toString
```
(also below for ranges...).

Reply:
resolve is not applicable here: resolve replaces the entire path of the uri. Storage namespace can include a path

Reply:
Not sure I understand:
```scala
scala> import java.net.URI
import java.net.URI

scala> val u = new URI("s3://a/bc/def")
val u: java.net.URI = s3://a/bc/def

scala> u.resolve("hij")
val res0: java.net.URI = s3://a/bc/hij
```

Reply:
OK, so I still think we can add a trailing slash to the repo.  The thing is, `URI.resolve` will avoid doubling slashes while still allowing e.g. passing a fully-qualified URL in the `Location` header.  URIs have standard semantics -- even though they are not nice, they are still useful.